### ISSUE TITLE
MSL Surface on animation window in OMEdit

### ISSUE THREAD

**Issue by BasilioPV** [OpenModelica/OpenModelica#8076]
### Description
Modelica.Mechanics.MultiBody.Visualizers.Advanced.Surface does not appear on animation window on OMEdit

### Steps to reproduce
Modelica.Mechanics.MultiBody.Examples.Elementary.Surfaces this example does not show any component that uses surface.

### Expected behavior
Surfaces must appear on animation window.

### Screenshots
Only wheel pipe part appears (not the surface external part). Surfaces, toroids, planes, etc.. does not appear

### Version and OS
 - OpenModelica Version 1.17.0 & 1.18.0
 - MSL 3.2.3 & 4.0.0
 - OS: Windows 10, 64 bit

**Comment by BasilioPV** [comment]
Any news on this issue?

**Comment by casella** [comment]
I'm not sure is @vwaurich is following, are you?

**Comment by adeas31** [comment]
@casella try reaching out @vwaurich via email.

**Comment by casella** [comment]
Done, let's see.

**Comment by vwaurich** [comment]
Surfaces need some more work to do. They rely on a replaceable function that calculates the positions of the vertices. See **Modelica.Mechanics.MultiBody.Interfaces.partialSurfaceCharacteristic()**
The current animation workflow needs either evaluated visualization parameters or variables that can be retrieved from the result file by name. A post-simulation evaluation of a function is not supported. An approach would be to compute the output vars of the surfaceCharacteristics-function during simulation and store it in the result files.

**Comment by BasilioPV** [comment]
Thanks vwaurich for your response. It is planned to include into result files the output vars of the surface Characteristics-function?

**Comment by casella** [comment]
@anotheruserofgithub we currently miss implementation of Surface and of Vector (see #8211), whose use was introduced in MSL 4.0.0. There is also some information in #8211. I will ask @perost to do all the frontend work for Surface as well in that ticket, then you could try to figure out how the current visualisers are implemented in OMEdit and extend them to support Surface (and also maybe Vector). At that point, the required skills are C/C++ programming and QT5, no need to worry about the OMC.

**Comment by perost** [comment]
> @perost Thank you for the quick commits! Do you think it would be possible in the meantime to go past `surfaceCharacteristic` and dump its `X`, `Y`, `Z` and `C` arrays instead? It seems to be the approach suggested by @vwaurich [in his comment above](https://github.com/OpenModelica/OpenModelica/issues/8076#issuecomment-987641561).

Yes, we could probably evaluate the functions and put the resulting arrays in the xml file instead of the functions themselves. That's still quite a lot of work though probably, so I don't know when I'll get around to looking into that. But if there's interest in it we might of course raise the priority.

**Comment by casella** [comment]
> @casella That's perfect! So I will look at that in my free time, first for Vector, then Surface.

Excellent, that would be great!

> I don't guarantee any results, though.

Of course. You just risk wasting your time 😄. If you need assistance on OMEdit programming you can ask @adeas31. You need to fork the https://github.com/OpenModelica/OpenModelica repo and eventually open a PR on master.
 
> @perost Thank you for the quick commits! Do you think it would be possible in the meantime to go past `surfaceCharacteristic` and dump its `X`, `Y`, `Z` and `C` arrays instead? It seems to be the approach suggested by @vwaurich [in his comment above](https://github.com/OpenModelica/OpenModelica/issues/8076#issuecomment-987641561).

**Comment by casella** [comment]
> Yes, we could probably evaluate the functions and put the resulting arrays in the xml file instead of the functions themselves. That's still quite a lot of work though probably, so I don't know when I'll get around to looking into that. But if there's interest in it we might of course raise the priority.

For the time being @perost's top priority is to help getting #8425 done. I'd suggest to start with Vector first.

**Comment by anotheruserofgithub** [comment]
I am almost done with implementing animation of the Surface visualizer in OMEdit, see [my WIP branch](https://github.com/anotheruserofgithub/OpenModelica/tree/surface_wip).
The one thing that is missing is to be able to get the vertices of the surface (`X`, `Y`, `Z`, `C` for colors, and `N` for normals if they are going to be introduced, see modelica/ModelicaStandardLibrary#4034) at any time frame, meaning, be able to call the surface characteristic function.
I am very much interested in getting this done soon.

I have been thinking about it and there are some constraints that (I think) were not considered previously.
By the way, tell me if I am wrong on any of the below points, as I am not an expert at all.
We could dump all the values directly in the results `_res.mat` file, but this could make it grow quite substantially, and since there is a size limit for `_res.mat` files as far as I understood, I do not think this is a good approach.
Nevertheless, it is not sufficient to just evaluate the replaceable function through the model instance API you are developing, because it would need to know the Modelica model.
While this is fine when simulating the model in OMEdit, it is not possible when the user only wants to open the `_res.mat` file (and the related `_visual.xml` file) in a new animation window, that is, without loading and resimulating the model.
Well, in fact, we could afford loading the model inside OMEdit automatically when opening the `_res.mat` file, assuming it is expected to exist in the same directory, but how fast would the model instance API be to evaluate the replaceable function(s)? Would it need to compile the whole model beforehand?
I personally have a use case with a lot of components and surfaces to visualize them, where compilation is a narrow bottleneck, and I wish I could just rerun the animation later without having to go through the whole model simulation pipeline again.
This would be useful also for demonstrations to other people, typically.

Hence, I think it may be better to compile all the surface characteristic functions of the model in a shared library that would exist alongside the `_res.mat` and `_visual.xml` files, and against which OMEdit could link dynamically.
Inside the `_visual.xml` file, for each surface a `<symbol>` tag would contain a string corresponding to a symbol of the shared library to be called in order to compute the vertices (this string could be constant because I do not think replaceable functions can be redeclared over time, can they?), something along these lines:
```xml
<surface>
    <ident>string</ident>
    <T>float array</T>
    <r>float array</r>
    <symbol>string</symbol>
    <nu>int</nu>
    <nv>int</nv>
    <wireframe>bool</wireframe>
    <normalized>bool</normalized>
    <doublesided>bool</doublesided>
    <multicolored>bool</multicolored>
    <color>float array</color>
    <specCoeff>float</specCoeff>
    <transparency>float</transparency>
</surface>
```
So the characteristic functions would be translated in C with the following prototype:
```c
void symbol(const double time,
            double* Vx, double* Vy, double* Vz,
            double* Nx, double* Ny, double* Nz,
            double* Cr, double* Cg, double* Cb);
```
They would fill in the output arrays passed as arguments, each of size `nu * nv`, which would have been allocated prior to the call (if the normals are not provided by the surface characteristic, `Nx/y/z` would be null thus left empty, and similarly if the surface is not multicolored, `Cr/g/b` would be null).
This shared library should be self-contained, i.e., by giving it only the current time stamp, it should be able to internally compute all the inputs required by the different implementations of the `partialSurfaceCharacteristic` interface function.

What do you think, would that be possible?
What issues do you see with such an approach?
Also, what if the surface characteristic has an external function interface, would that work correctly?
Or if it calls other (non-)built-in functions?
And what about event generation and other constructs that are allowed in (replaceable) functions by the standard?

I assume you already have tools to generate C code from Modelica code, but I do not know if they could be used straightforwardly or would need modifications.
Furthermore, @perost was saying that replaceable functions are not handled at all yet (?) since they are not part of the flat model, so what should be done to make use of them?

I would appreciate any feedbacks!

**Comment by casella** [comment]
@anotheruserofgithub, @adeas31 and @perost are off until Jan 9, I'm afraid you'll have to wait until then to get some feedback. 

> Furthermore, @perost was saying that replaceable functions are not handled at all yet (?) since they are not part of the flat model, so what should be done to make use of them?

I'm not sure I understand this. The flat model is obtained by applying all the redeclares. Of course this is done at compile time, then they are hardwired.

**Comment by anotheruserofgithub** [comment]
Thanks for the info!

> I'm not sure I understand this.

I was referring to him saying

> The function isn't even part of the flat model since it's not called anywhere in the model, so we'll have to somehow instantiate it and pass it to the backend in some way.

in https://github.com/OpenModelica/OpenModelica/issues/8211#issuecomment-1134512540 but surely I misunderstood, I guess it is just that the characteristic function is dropped since it is unused, right?

**Comment by casella** [comment]
> in [#8211 (comment)](https://github.com/OpenModelica/OpenModelica/issues/8211#issuecomment-1134512540) but surely I misunderstood, I guess it is just that the characteristic function is dropped since it is unused, right?

Unfortunately I'm not too familiar with how animations are handled. It looks like this function is needed by the GUI for rendering but is not really called by the Modelica model, so it is dropped by the frontend.

**Comment by perost** [comment]
> in [#8211 (comment)](https://github.com/OpenModelica/OpenModelica/issues/8211#issuecomment-1134512540) but surely I misunderstood, I guess it is just that the characteristic function is dropped since it is unused, right?

Yes, the fact that it's replaceable isn't an issue, the issue is that there's no practical way to add the functions to the XML we currently use. Your proposed solution might be possible, but I'm not exactly an expert on these things either unfortunately.

But it seems to me like implementing it properly might require a non-trivial amount of work, so we might need to discuss it a bit in our team and possibly allocate some resources to it.

**Comment by onesandineye** [comment]
Any news about this issue?

**Comment by casella** [comment]
@adeas31 can give a look when he gets back from his vacation.

**Comment by anotheruserofgithub** [comment]
Hi @onesandineye, I already worked a lot on this stuff two years ago but I had to put it on hold. I will get back to it after I finish https://github.com/OpenModelica/OpenModelica/issues/9390#issuecomment-3145843548. Hopefully it will be available in the next release of OM.
